packages/server/src/fhir/repo.ts

```diff
@@ -1668,90 +1671,68 @@ export class Repository extends FhirRepository<PoolClient> implements Disposable
    * If the search parameter is not an array, then this method will be called for the value.
    * @param searchParam - The search parameter definition.
    * @param details - The extra search parameter details.
-   * @param value - The FHIR resource value.
+   * @param typedValues - The FHIR resource value.
    * @returns The column value.
    */
-  private buildColumnValue(searchParam: SearchParameter, details: SearchParameterDetails, value: any): any {
+  private buildColumnValues(
+    searchParam: SearchParameter,
+    details: SearchParameterDetails,
+    typedValues: TypedValue[]
+  ): (boolean | number | string | undefined)[] {
     if (details.type === SearchParameterType.BOOLEAN) {
-      return value === true || value === 'true';
+      const value = typedValues[0]?.value;
+      return [value === true || value === 'true'];
     }
 
     if (details.type === SearchParameterType.DATE) {
-      return this.buildDateColumn(value);
+      // "Date" column is a special case that only applies when the following conditions are true:
+      // 1. The search parameter is a date type.
+      // 2. The underlying FHIR ElementDefinition referred to by the search parameter has a type of "date".
+      return convertToSearchableDates(typedValues)
+        .map((p) => (p.start ?? p.end)?.substring(0, 10))
+        .filter(Boolean) as string[];
     }
 
     if (details.type === SearchParameterType.DATETIME) {
-      return this.buildDateTimeColumn(value);
+      return convertToSearchableDates(typedValues)
+        .map((p) => p.start ?? p.end)
+        .filter(Boolean) as string[];
+    }
+
+    if (searchParam.type === 'number') {
+      return convertToSearchableNumbers(typedValues).filter((n) => n !== undefined);
     }
 
     if (searchParam.type === 'quantity') {
-      return this.buildQuantityColumn(value);
+      return convertToSearchableQuantities(typedValues)
+        .map((q) => q.value)
+        .filter((q) => q !== undefined);
```

Comment:
Probably not a big deal, but these could use our `flatMapFilter()` helper to perform a single loop over the initial array and create only a single output array